### PR TITLE
switch to @replit/node-fetch if they try to import node-fetch 3 in cjs

### PR DESCRIPTION
**Pull request by airportyh**
# Why

https://app.asana.com/0/1186764502771137/1202131546027041

## What changed?

Added `Module._findPath` monkey patch to check for conditions:

1. the code is in commonjs mode (does not have `.mjs` extension)
2. we are requesting module node-fetch
3. after resolving node-fetch the normal way, we find that its version is 3 or above

If all 3 are true, we switch the resolution to `@replit/node-fetch` module, which is bundled with prybar-nodejs.

Tried using yarn to install `@replit/node-fetch` in build time, but it had a weird EPERM error when copying over files which I never got to the bottom of. Instead just using npm in the build steps now.

## Test

Run the tests.

### CODE REVIEW

**Comment by alex-4ed5ec60d** [comment]
Good news! I’m the author of `node-fetch-commonjs`, so if it doesn’t exist you can just ping me on Slack.

**Comment by alex-4ed5ec60d** [comment]
That being said, please fork it to `@replit/node-fetch` so it looks more official and people would be less concerned about unwillingly installing a package from an “unknown author”.

**Comment by airportyh** [comment]
> Good news! I’m the author of `node-fetch-commonjs`, so if it doesn’t exist you can just ping me on Slack.

Haha! I just meant if we don't have it installed in the repl.

File: prybar_assets/nodejs/module-context-hook.js [hunk 1]
```diff
@@ -9,6 +9,28 @@ const nodeRepl = require('repl');
...
+    const dir = path.dirname(result);
+    const packageJson = require(path.join(dir, "..", "package.json"));
+    const [major] = packageJson.version.split(".").map(Number);
+    if (major >= 3) {
+      // Switch to node-fetch-commonjs
+      console.log("\u001b[33mAuto-switching node-fetch to node-fetch-commonjs\u001b[0m");
+      const newResult = result.replace('node_modules/node-fetch/src', 'node_modules/node-fetch-commonjs');
+      return newResult;
+    }
+    return result;
```
Comment (AllAwesome497) [inline prybar_assets/nodejs/module-context-hook.js:29]: nit: add empty lines before `return` (except at the start of blocks). 
```suggestion
      const newResult = result.replace('node_modules/node-fetch/src', 'node_modules/node-fetch-commonjs');

      return newResult;
    }

    return result;
```

File: prybar_assets/nodejs/module-context-hook.js [hunk 2]
```diff
@@ -9,6 +9,28 @@ const nodeRepl = require('repl');
...
+    // Because node-fetch version >= 3 doesn't support commonjs, we'll switch to
+    // node-fetch-commonjs if it comes to that
+    const result = originalFindPath(request, paths, isMain);
+    const dir = path.dirname(result);
+    const packageJson = require(path.join(dir, "..", "package.json"));
+    const [major] = packageJson.version.split(".").map(Number);
+    if (major >= 3) {
+      // Switch to node-fetch-commonjs
+      console.log("\u001b[33mAuto-switching node-fetch to node-fetch-commonjs\u001b[0m");
+      const newResult = result.replace('node_modules/node-fetch/src', 'node_modules/node-fetch-commonjs');
```
Comment (AllAwesome497) [inline prybar_assets/nodejs/module-context-hook.js:26]: we should bundle `node-fetch-commonjs`.  UPM won't automatically install this package.

**Review by AllAwesome497 — CHANGES_REQUESTED** [review]
looks good, we just need to make sure that we bundle `node-fetch-commonjs` so this doesn't break.

File: prybar_assets/nodejs/module-context-hook.js (hunk 2 above)
Comment (masad-frost) [reply prybar_assets/nodejs/module-context-hook.js:26]: +1, we should set the path to the asset-dir, https://github.com/replit/prybar/blob/a584d8e52bbff356140546fc6691f1d2c01a176e/languages/nodejs/repl.js#L3-L8

Comment (alex-4ed5ec60d) [inline prybar_assets/nodejs/module-context-hook.js:26]: `@replit/node-fetch`

File: .gitignore [hunk 3]
```diff
@@ -5,3 +5,5 @@ deps.lst
 prybar.tar.gz
 generated_*.go
 result*
+prybar_assets/nodejs/node_modules
+test_files/node_modules
```
Comment (AllAwesome497) [inline .gitignore:9]:
```suggestion
test_files/node_modules

```

File: languages/nodejs/compile [hunk 4]
```diff
@@ -0,0 +1,8 @@
+#!/bin/sh
+
+set -e
+
+NODEJS_LANGUAGE_DIR="$(realpath $(dirname $0))"
+PRYBAR_ROOT_DIR="$(dirname $(dirname "${NODEJS_LANGUAGE_DIR}"))"
+
+cd "${PRYBAR_ROOT_DIR}/prybar_assets/nodejs" && HOME="" npm install
```
Comment (AllAwesome497) [inline languages/nodejs/compile:8]:
```suggestion
cd "${PRYBAR_ROOT_DIR}/prybar_assets/nodejs" && HOME="" npm install

```

File: prybar_assets/nodejs/module-context-hook.js [hunk 5]
```diff
@@ -9,6 +9,32 @@ const nodeRepl = require('repl');
...
+
+      return path.join(assets_dir, "nodejs/node_modules/@replit/node-fetch/index.js");
+    }
+
+    return result;
+  } else {
+    
+    return originalFindPath(request, paths, isMain);
+  }
+}
```
Comment (AllAwesome497) [inline prybar_assets/nodejs/module-context-hook.js:37]: this should probably be in `repl.js`.  This file is specifically for the module context magic which allows the console to reference variables in module scope.

File: test_files/package.json [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+  "name": "test_files",
+  "version": "1.0.0",
+  "description": "",
+  "main": "global_bug.js",
+  "scripts": {
+    "test": "echo \"Error: no test specified\" && exit 1"
+  },
+  "keywords": [],
+  "author": "",
+  "license": "ISC",
```
Comment (AllAwesome497) [inline test_files/package.json:11]: ISC?  what's the tl;dr there?

File: test_files/require_node_fetch.js [hunk 7]
```diff
@@ -0,0 +1,3 @@
+const fetch = require('node-fetch');
+
+console.log('ok');
```
Comment (AllAwesome497) [inline test_files/require_node_fetch.js:3]:
```suggestion
console.log('ok');

```

File: tests/nodejs/node_fetch.exp [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+spawn sh
+send "export PRYBAR_ASSETS_DIR=\$PWD/prybar_assets\r"
+send "cd test_files\r"
+send "../prybar-nodejs -i require_node_fetch.js\r"
+match_max 100000
+expect -re {Node v\d+\.\d+.\d+ on (linux|darwin)}
+expect -re {Auto\-switching}
+expect -re {ok}
+expect -exact "\[0m\[90mHint: hit control+c anytime to enter REPL.\[0m\r
+\[1G\[0J--> \[5G"
```
Comment (AllAwesome497) [inline tests/nodejs/node_fetch.exp:14]:
```suggestion
\[1G\[0J--> \[5G"

```

**Review by AllAwesome497 — APPROVED** [review]
looks fine to me

File: test_files/package.json (hunk 6 above)
Comment (airportyh) [reply test_files/package.json]: Will remove it.

File: prybar_assets/nodejs/module-context-hook.js (hunk 5 above)
Comment (airportyh) [reply prybar_assets/nodejs/module-context-hook.js:37]: I am using context within runModule that knows whether the file is run in cjs vs esm mode. Not very convenient to move out for that reason.

**Pull request merged**